pkg/components/bindings/registry.go

```diff
@@ -80,33 +80,55 @@ func (b *bindingsRegistry) RegisterOutputBindings(components ...OutputBinding) {
 }
 
 // Create instantiates an input binding based on `name`.
-func (b *bindingsRegistry) CreateInputBinding(name string) (bindings.InputBinding, error) {
-	if method, ok := b.inputBindings[name]; ok {
+func (b *bindingsRegistry) CreateInputBinding(name, version string) (bindings.InputBinding, error) {
+	if method, ok := b.getInputBinding(name, version); ok {
 		return method(), nil
 	}
-	return nil, errors.Errorf("couldn't find input binding %s", name)
+	return nil, errors.Errorf("couldn't find input binding %s/%s", name, version)
 }
 
 // Create instantiates an output binding based on `name`.
-func (b *bindingsRegistry) CreateOutputBinding(name string) (bindings.OutputBinding, error) {
-	if method, ok := b.outputBindings[name]; ok {
+func (b *bindingsRegistry) CreateOutputBinding(name, version string) (bindings.OutputBinding, error) {
+	if method, ok := b.getOutputBinding(name, version); ok {
 		return method(), nil
 	}
-	return nil, errors.Errorf("couldn't find output binding %s", name)
+	return nil, errors.Errorf("couldn't find output binding %s/%s", name, version)
 }
 
 // HasInputBinding checks if an input binding based on `name` exists in the registry.
-func (b *bindingsRegistry) HasInputBinding(name string) bool {
-	_, ok := b.inputBindings[name]
+func (b *bindingsRegistry) HasInputBinding(name, version string) bool {
+	_, ok := b.getInputBinding(name, version)
 	return ok
 }
 
 // HasOutputBinding checks if an output binding based on `name` exists in the registry.
-func (b *bindingsRegistry) HasOutputBinding(name string) bool {
-	_, ok := b.outputBindings[name]
+func (b *bindingsRegistry) HasOutputBinding(name, version string) bool {
+	_, ok := b.getOutputBinding(name, version)
 	return ok
 }
 
+func (b *bindingsRegistry) getInputBinding(name, version string) (func() bindings.InputBinding, bool) {
+	binding, ok := b.inputBindings[name+"/"+version]
+	if ok {
+		return binding, true
+	}
+	if version == "" || version == "v0" || version == "v1" {
```

Comment:
It is odd that v0 and v1 has the same treatment. Maybe we should just handle empty. This brings to another discussion. What is the meaning of version? If there is a patch, the version will not change since it would just be a fix in the code. A minor version change is unlikely because it means that it would be a feature addition. It feels like the only time version would change (new version added to the list) is for a major version. It would be much simpler if this was just an integer, IMO.

Reply:
According to Go semantics, v0 and v1 do not need a special package. only starts with v2+.

As per our versioning guidelines, a non-breaking change/patch does not increment the version.

Reply:
> It is odd that v0 and v1 has the same treatment. Maybe we should just handle empty.

The thought here is that we _could_ use the same versioning approach as [Go modules](https://blog.golang.org/v2-go-modules) where the unstable version (v0) and first stable version (v1) are contained in the same package and v2 and beyond are in a sub-package. See dapr/components-contrib#650 for the accompanying documentation. `"v0" || "v1"` are to redirect the lookup from the deployment YAML to how the components are currently registered in Dapr's `main.go` where there are currently no version suffixes.

The fact that v0 and v1 have the same treatment could be viewed as backward compatibility for registration. In the future, we might require that components be explicitly registered with a version (separate registrations for v0 and v1). Or we could decide to completely embrace the Go modules approach where no suffix follows v0 through v1 of a components lifecycle. I went with the later option for now as it seems to fit Go's versioning model.

> What is the meaning of version? If there is a patch, the version will not change since it would just be a fix in the code. A minor version change is unlikely because it means that it would be a feature addition. It feels like the only time version would change (new version added to the list) is for a major version.

You are correct. Version is really the major version per the Go module approach above and also explained in dapr/components-contrib#650; hopefully providing the right guidance for _when_ and _how_ for create version 2+.

> It would be much simpler if this was just an integer, IMO.

Totally agree in that since version indicates major version, it could be passed as an integer. It is currently a string because it comes from the deployment YAML which contains values with values like `v1`, `v2`, etc..

Reply:
OK, so we are following Go semantics. Then, can we convert the value from string to integer at the earliest point? Then the whole code will be much simpler (contains "v"?, upper case or lower case?, trailing spaces?) and we can easily have a notion of order too.

Reply:
Sadly, because the `ComponentSpec` struct is passed around extensively inside `runtime.go`, it would be difficult to address it at an early point. The code would a bit more complex because the conversion would happen at each point where a `registry.Create` method is called. To address this properly, we should consider changing the `Version` field of  `ComponentSpec` to an integer. But at this point `v1` values are prevalent in the codebase and configurations so my sense is that change would be outside the scope of this PR.

The string value may be offer flexibility but also could introduce human error. I'm hoping the docs I've added in dapr/components-contrib at least will minimize the potential for misconfiguration. I may add case-insensitivity to the registries to at least protect against `v1` vs `V1`.

Reply:
I don't support changing the Version field to be an integer, as it might limit the introduction of new API groups later on.

I think it's reasonable to add case-insensitivity.

Reply:
Case-insensitivity added. See https://github.com/dapr/dapr/pull/2757/commits/55bd4dcdc16ec2ce35a722258e3b56d5b445fce5.